### ISSUE TITLE
Support query string in edit_uri

### ISSUE THREAD

**Issue by ovasquez** [mkdocs/mkdocs#1220]
The `edit_uri` setting in the `mkdocs.yml` file doesn't support the question mark, which is a valid and very common character in URLs. When a '?' is used the edit_uri contents are ignored and the edit URL is incorrectly generated.

This approach is used in the _Visual Studio Team Services_ git repository, so it would be really helpful to be able to link the files.
For example: 
```
edit_uri: '?path=/website/docs/'
```
This also happens if the '?' is added in the `repo_url` setting.

**Comment by waylan** [comment]
This appears to be the behavior of `urljoin`:

>>> urljoin('http://example.com/foo/bar', 'baz/blah.txt')
'http://example.com/foo/baz/blah.txt'
>>> urljoin('http://example.com/?path=/foo/bar', 'baz/blah.txt')
'http://example.com/baz/blah.txt'

The first example is what we anticipated. The first string points to the location and the second string is the path from the document root to the file. But when the first string includes a query string (part that starts with `?`) then is strips that part out as we see in the second example. This actually makes sense. It would be wrong to have a query string in the middle of a URL and `joinurl` wouldn't necessarily consider that you might want to join a string to the query string. Consider this example:

urljoin('http://example.com/', '?path=/foo/bar', 'baz/blah.txt')
'http://example.com/?path=/foo/bar'

There it adds the query string, but then ignores everything after it. Looking at the [source], we don't use `joinurl` to join the repo_url and the edit_url, but we do use joinurl to join the file path to the combined repo_url/edit_url.

[source]: https://github.com/mkdocs/mkdocs/blob/b3ca24234b5cbd29d401dd8477fe0d9a7c8f48a4/mkdocs/nav.py#L214-L216

**Comment by ovasquez** [comment]
@waylan I played around with the code and noticed that the *query string* is supported in the second parameter of *urljoin*.
Is there any case where changing this code in the source could create an issue? From this:
```
self.edit_url = utils.urljoin(
                repo_url + edit_uri,
                self.input_path)
```
to this:
```
self.edit_url = utils.urljoin(
                repo_url,
                edit_uri + self.input_path)
```
Based on the usage of the `edit_uri` it should be the root and if I understand correctly `self.input_path` is relative so there should be no overlap between them, or am I missing something?

**Comment by waylan** [comment]
@ovasquez maybe that would work. I'm not sure yet. Notice [a few lines of code before][1] the subject lines. There we check that `repo_url` ends with a `/` (but only one) so that when we concatenate the strings we get a proper URL. Normally that would be taken care of by `urljoin` but we do it ourselves because we need to join three parts. We could do the same for `edit_uri` and `self.input_path`, but `urljoin` also does some extra stuff when joining two URLs which a common ancestor that we would loose. I realize that is not relevant to your situation, but it could be to others. My point is, it may not be that simple. On the other hand, maybe those concerns don't exist and it is that simple (in which case we don't need to use `urljoin` at all). That needs to be explored. Until we have a definitive answer, we don't have a fix.

Of course, a PR with a fix that included tests of various scenarios would be welcome.

[1]: https://github.com/mkdocs/mkdocs/blob/b3ca24234b5cbd29d401dd8477fe0d9a7c8f48a4/mkdocs/nav.py#L207

**Comment by ovasquez** [comment]
@waylan Makes sense. I might be able to write some tests eventually, but I have one more question.
The are three main variables involving the solution:
- `repo_url`: absolute URL to the repo
- `edit_uri`: relative URL
- `self.input_path`: From what I see in the source code and tests it's just the normalized path from the mkdocs.yml *page* configuration.

Is there any way to find out which could be all the expected formats of `self.input_path`? Or the example below is enough to cover the necessary cases?

`self.input_path` formats example:
```
pages = [
            'index.md',
            'internal.md',
            'sub/internal.md',
            'sub1/sub2/internal.md',
        ]
site_navigation = nav.SiteNavigation(pages)
...
```

**Comment by waylan** [comment]
Yeah those look right to me.

**Comment by ovasquez** [comment]
@waylan I believe this issue could be closed with the pull request #1224.
Any idea of when the updated python package will be released?

**Comment by waylan** [comment]
Ahh, I always include a reference to the open issue in my commit messages, which causes the issue to be closed automatically when merged. I missed that you didn't do that.
